### PR TITLE
magic-wormhole: fix build

### LINKED ISSUE #309210: Build failure: magic-wormhole 0.14.0
### Steps To Reproduce

1. build magic-wormhole on master

### Build log

```
error: builder for '/nix/store/8jnqy1jfbrwag1v12abndydy4rf49jp6-python3.11-magic-wormhole-0.14.0.drv' failed with exit code 1;
       last 10 log lines:
       > adding 'magic_wormhole-0.14.0.dist-info/RECORD'
       > removing build/bdist.linux-x86_64/wheel
       > Successfully built magic_wormhole-0.14.0-py3-none-any.whl
       > Finished creating a wheel...
       > Finished executing pypaBuildPhase
       > Running phase: pythonRuntimeDepsCheckHook
       > Executing pythonRuntimeDepsCheck
       > Checking runtime dependencies for magic_wormhole-0.14.0-py3-none-any.whl
       >   - zipstream-ng not installed
       >   - iterable-io not installed
       For full logs, run 'nix log /nix/store/8jnqy1jfbrwag1v12abndydy4rf49jp6-python3.11-magic-wormhole-0.14.0.drv'.
```

### Additional context

The two dependencies were in [this upstream commit](https://github.com/magic-wormhole/magic-wormhole/commit/c4c56e1d75a3f25246d785f3635c67a81ea0193d). I think iterable-io is not packaged in Nix yet?

### Notify maintainers

@asymmetric @mweinelt

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.7, NixOS, 24.05 (Uakari), 24.05.20240423.572af61`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/m5i890m2g4pnyflpn48d1dpzzmwp5q4p-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by mjoerg**
## Description of changes

This adds the missing dependencies required for magic-wormhole 0.14.0, and also improves dependencies
magic-wormhole-mailbox-server and magic-wormhole-transit-relay.

Fixes #309210.

## Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>gnome-keysign</li>
    <li>gnome-keysign.dist</li>
    <li>magic-wormhole (python311Packages.magic-wormhole)</li>
    <li>magic-wormhole.dist (python311Packages.magic-wormhole.dist)</li>
    <li>python311Packages.iterable-io</li>
    <li>python311Packages.iterable-io.dist</li>
    <li>python312Packages.iterable-io</li>
    <li>python312Packages.iterable-io.dist</li>
    <li>tahoe-lafs</li>
    <li>tahoe-lafs.dist</li>
    <li>tahoe-lafs.doc</li>
    <li>tahoe-lafs.info</li>
  </ul>
</details>

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/magic-wormhole-mailbox-server/default.nix [hunk 1]
```diff
@@ -19,11 +18,12 @@ buildPythonPackage rec {
   version = "0.4.1";
   format = "setuptools";
 
-  disabled = pythonOlder "3.7";
+  # python 3.12 support: https://github.com/magic-wormhole/magic-wormhole-mailbox-server/issues/41
+  disabled = isPy312 || pythonOlder "3.7";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/magic-wormhole-mailbox-server/default.nix:22]:
```suggestion
  disabled = pythonOlder "3.7" || pythonAtLeast "3.12";
```

File: pkgs/development/python-modules/magic-wormhole-transit-relay/default.nix [hunk 2]
```diff
@@ -4,16 +4,20 @@
...
+, isPy312
+, pythonOlder
 }:
 
 buildPythonPackage rec {
   pname = "magic-wormhole-transit-relay";
   version = "0.2.1";
   format = "setuptools";
 
+  disabled = isPy312 || pythonOlder "3.7";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/magic-wormhole-transit-relay/default.nix:16]:
```suggestion
  disabled = pythonOlder "3.7" || pythonAtLeast "3.12";
```

File: pkgs/development/python-modules/magic-wormhole-transit-relay/default.nix [hunk 3]
```diff
@@ -4,16 +4,20 @@
...
 , mock
 , twisted
+, isPy312
+, pythonOlder
 }:
 
 buildPythonPackage rec {
   pname = "magic-wormhole-transit-relay";
   version = "0.2.1";
   format = "setuptools";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/magic-wormhole-transit-relay/default.nix:14]: Can we switch that to pyproject = true?

**Review by SuperSandro2000 — COMMENTED** [review]
You want to adopt the other packages that have no maintainer, too?

File: pkgs/development/python-modules/magic-wormhole/default.nix [hunk 4]
```diff
@@ -111,12 +115,12 @@ buildPythonPackage rec {
...
 
-  meta = with lib; {
+  meta = {
     changelog = "https://github.com/magic-wormhole/magic-wormhole/blob/${version}/NEWS.md";
     description = "Securely transfer data between computers";
     homepage = "https://github.com/magic-wormhole/magic-wormhole";
-    license = licenses.mit;
-    maintainers = with maintainers; [ asymmetric ];
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ asymmetric ];
```
Comment (asymmetric) [inline pkgs/development/python-modules/magic-wormhole/default.nix:123]: @mjoerg do you want to maintain this as well? I don't use it anymore, I've switched to the Rust version when I need something like this.

**Review by asymmetric — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**